File: whdon_app/src/main/java/nl/whitedove/washetdroogofniet/StatsWeerTypeActivity.java
package nl.whitedove.washetdroogofniet;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.view.View;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.components.Description;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.LegendEntry;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;

import org.joda.time.DateTime;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Locale;

public class StatsWeerTypeActivity extends Activity {
    static int mJaar = DateTime.now().getYear();
    static int mMaand = DateTime.now().getMonthOfYear();
    static Helper.Periode mAllesJaarMaand = Helper.Periode.Maand;

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.weertype_statistieken);
        InitFab();
        InitRadio();
        InitSwipes();
        ToondataBackground();
    }

    private void InitFab() {
        FloatingActionButton fabTerug = findViewById(R.id.btnTerug);
        fabTerug.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Terug();
            }
        });
    }

    private void InitRadio() {
        final RadioButton rbAlles = findViewById(R.id.rbAlles);
        final RadioButton rbJaar = findViewById(R.id.rbJaar);
        final RadioButton rbMaand = findViewById(R.id.rbMaand);
        final RadioGroup rgAllesJaarMaand = findViewById(R.id.rgAllesJaarMaand);
        rbAlles.setChecked(mAllesJaarMaand == Helper.Periode.Alles);
        rbJaar.setChecked(mAllesJaarMaand == Helper.Periode.Jaar);
        rbMaand.setChecked(mAllesJaarMaand == Helper.Periode.Maand);
        RadioGroup.OnCheckedChangeListener cl = new RadioGroup.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(RadioGroup radioGroup, int checkedId) {
                RadioButton rb = radioGroup.findViewById(checkedId);
                if (rb.getId() == R.id.rbAlles) {
                    mAllesJaarMaand = Helper.Periode.Alles;
                }

                if (rb.getId() == R.id.rbJaar) {
                    mAllesJaarMaand = Helper.Periode.Jaar;
                    mJaar = DateTime.now().getYear();
                }

                if (rb.getId() == R.id.rbMaand) {
                    mAllesJaarMaand = Helper.Periode.Maand;
                    mMaand = DateTime.now().getMonthOfYear();
                }
                ToondataBackground();
            }
        };
        rgAllesJaarMaand.setOnCheckedChangeListener(cl);
    }

    @SuppressLint("ClickableViewAccessibility")
    private void InitSwipes() {
        OnSwipeTouchListener sl = new OnSwipeTouchListener(StatsWeerTypeActivity.this) {
            public void onSwipeLeft() {
                if (mAllesJaarMaand == Helper.Periode.Jaar) {
                    mJaar++;
                }

                if (mAllesJaarMaand == Helper.Periode.Maand) {
                    mMaand++;
                    if (mMaand == 13) {
                        mJaar++;
                        mMaand = 1;
                    }
                }
                ToondataBackground();
            }

            public void onSwipeRight() {
                if (mAllesJaarMaand == Helper.Periode.Jaar) {
                    mJaar--;
                }

                if (mAllesJaarMaand == Helper.Periode.Maand) {
                    mMaand--;
                    if (mMaand == 0) {
                        mJaar--;
                        mMaand = 12;
                    }
                }
                ToondataBackground();
            }
        };

        final RelativeLayout rlPerMaand = findViewById(R.id.rlPerWeertype);
        rlPerMaand.setOnTouchListener(sl);

        final PieChart pChart = findViewById(R.id.pcPerWeertype);
        pChart.setOnTouchListener(sl);

        Helper.INSTANCE.showMessage(StatsWeerTypeActivity.this, getString(R.string.SwipeLinksOfRechts));
    }

    private void Terug() {
        Intent intent = new Intent(this, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
    }

    private void ToondataBackground() {
        Context context = getApplicationContext();
        new AsyncGetStatistiekWeerTypeTask(this).execute(context);
    }

    @SuppressLint("DefaultLocale")
    private void ToonStatistiekWeerType(ArrayList<StatistiekWeertype> stats) {

        final TextView tvWeertype = findViewById(R.id.tvWeertype);
        final TextView tvGeenGegevens = findViewById(R.id.tvGeenGegevens);

        if (mAllesJaarMaand == Helper.Periode.Alles) {
            tvWeertype.setText(String.format("%s", getString(R.string.PerWeerType)));
        }

        if (mAllesJaarMaand == Helper.Periode.Jaar) {
            tvWeertype.setText(String.format("%s %s", getString(R.string.PerWeerType), String.format(getString(R.string.Jaartal), Integer.toString(mJaar))));
        }

        if (mAllesJaarMaand == Helper.Periode.Maand) {
            DateTime dat = new DateTime(mJaar, mMaand, 1, 0, 0);
            String mnd = dat.toString("MMM", Locale.getDefault()).replace(".", "");
            tvWeertype.setText(String.format("%s %s", getString(R.string.PerWeerType), String.format(getString(R.string.JaartalEnMaand), mnd, Integer.toString(mJaar))));
        }

        ArrayList<Integer> colors = new ArrayList<>();
        ArrayList<PieEntry> dataT = new ArrayList<>();
        ArrayList<LegendEntry> legendEntries = new ArrayList<>();

        for (int i = 0; i < stats.size(); i++) {
            float perc = stats.get(i).getPercentage();
            int rPerc = Math.round(perc);
            dataT.add(new PieEntry(perc, (rPerc <= 2) ? "" : String.format("%d%%", rPerc)));
            LegendEntry le = new LegendEntry();
            Integer col = WeerHelper.WeerTypeToWeerKleur(this, stats.get(i).getWeerType());
            colors.add(col);
            //noinspection ConstantConditions
            le.formColor = col;
            le.label = stats.get(i).getWeerTypeOmschrijving();
            le.form = Legend.LegendForm.SQUARE;
            le.formSize = 10;
            legendEntries.add(le);
        }

        final PieChart chart = findViewById(R.id.pcPerWeertype);

        Description desc = new Description();
        desc.setText("");
        chart.setDescription(desc);

        chart.setTouchEnabled(false);
        chart.setNoDataText(getString(R.string.nodata));
        chart.getLegend().setEnabled(false);
        chart.setEntryLabelColor(ContextCompat.getColor(this, R.color.colorPrimaryDark));
        chart.setDrawEntryLabels(true);

        Legend legend = chart.getLegend();
        legend.setCustom(legendEntries);
        legend.setEnabled(true);
        legend.setXEntrySpace(20f);
        legend.setTextSize(12f);
        legend.setWordWrapEnabled(true);

        PieDataSet dsT = new PieDataSet(dataT, "");
        dsT.setColors(colors);
        dsT.setDrawValues(false);

        PieData data = new PieData(dsT);
        data.setValueTextSize(14f);
        data.setValueTextColor(ContextCompat.getColor(this, R.color.colorTekst));

        chart.setData(data);
        chart.animateXY(500, 500);
        if (stats.size() == 0) {
            chart.setVisibility(View.GONE);
            tvGeenGegevens.setVisibility(View.VISIBLE);
        } else {
            tvGeenGegevens.setVisibility(View.GONE);
            chart.setVisibility(View.VISIBLE);
        }
        chart.invalidate();

    }

    private static class AsyncGetStatistiekWeerTypeTask extends AsyncTask<Context, Void, ArrayList<StatistiekWeertype>> {
        private WeakReference<StatsWeerTypeActivity> activityWeakReference;

        AsyncGetStatistiekWeerTypeTask(StatsWeerTypeActivity context) {
            activityWeakReference = new WeakReference<>(context);
        }

        @Override
        protected ArrayList<StatistiekWeertype> doInBackground(Context... params) {
            Context context = params[0];
            DatabaseHelper dh = DatabaseHelper.Companion.getInstance(context);
            return dh.GetStatistiekWeerType(mAllesJaarMaand, mJaar, mMaand);
        }

        @Override
        protected void onPostExecute(ArrayList<StatistiekWeertype> stats) {
            StatsWeerTypeActivity activity = activityWeakReference.get();
            if (activity != null) activity.ToonStatistiekWeerType(stats);
        }
    }
}

File: whdon_app/src/main/java/nl/whitedove/washetdroogofniet/StatsWeerTypeActivity.kt
package nl.whitedove.washetdroogofniet

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import com.google.android.material.floatingactionbutton.FloatingActionButton
import androidx.core.content.ContextCompat
import android.view.View
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.RelativeLayout
import android.widget.TextView

import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.components.Description
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.components.LegendEntry
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry

import org.joda.time.DateTime

import java.lang.ref.WeakReference
import java.util.ArrayList
import java.util.Locale

class StatsWeerTypeActivity : Activity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.weertype_statistieken)
        initFab()
        initRadio()
        initSwipes()
        toondataBackground()
    }

    private fun initFab() {
        val fabTerug = findViewById<FloatingActionButton>(R.id.btnTerug)
        fabTerug.setOnClickListener { terug() }
    }

    private fun initRadio() {
        val rbAlles = findViewById<RadioButton>(R.id.rbAlles)
        val rbJaar = findViewById<RadioButton>(R.id.rbJaar)
        val rbMaand = findViewById<RadioButton>(R.id.rbMaand)
        val rgAllesJaarMaand = findViewById<RadioGroup>(R.id.rgAllesJaarMaand)
        rbAlles.isChecked = mAllesJaarMaand === Helper.Periode.Alles
        rbJaar.isChecked = mAllesJaarMaand === Helper.Periode.Jaar
        rbMaand.isChecked = mAllesJaarMaand === Helper.Periode.Maand
        val cl = RadioGroup.OnCheckedChangeListener { radioGroup, checkedId ->
            val rb = radioGroup.findViewById<RadioButton>(checkedId)
            if (rb.id == R.id.rbAlles) {
                mAllesJaarMaand = Helper.Periode.Alles
            }

            if (rb.id == R.id.rbJaar) {
                mAllesJaarMaand = Helper.Periode.Jaar
                mJaar = DateTime.now().year
            }

            if (rb.id == R.id.rbMaand) {
                mAllesJaarMaand = Helper.Periode.Maand
                mMaand = DateTime.now().monthOfYear
            }
            toondataBackground()
        }
        rgAllesJaarMaand.setOnCheckedChangeListener(cl)
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun initSwipes() {
        val sl = object : OnSwipeTouchListener(this@StatsWeerTypeActivity) {
            override fun onSwipeLeft() {
                if (mAllesJaarMaand === Helper.Periode.Jaar) {
                    mJaar++
                }

                if (mAllesJaarMaand === Helper.Periode.Maand) {
                    mMaand++
                    if (mMaand == 13) {
                        mJaar++
                        mMaand = 1
                    }
                }
                toondataBackground()
            }

            override fun onSwipeRight() {
                if (mAllesJaarMaand === Helper.Periode.Jaar) {
                    mJaar--
                }

                if (mAllesJaarMaand === Helper.Periode.Maand) {
                    mMaand--
                    if (mMaand == 0) {
                        mJaar--
                        mMaand = 12
                    }
                }
                toondataBackground()
            }
        }

        val rlPerMaand = findViewById<RelativeLayout>(R.id.rlPerWeertype)
        rlPerMaand.setOnTouchListener(sl)

        val pChart = findViewById<PieChart>(R.id.pcPerWeertype)
        pChart.setOnTouchListener(sl)

        Helper.showMessage(this@StatsWeerTypeActivity, getString(R.string.SwipeLinksOfRechts))
    }

    private fun terug() {
        val intent = Intent(this, MainActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
        startActivity(intent)
    }

    private fun toondataBackground() {
        val context = applicationContext
        AsyncGetStatistiekWeerTypeTask(this).execute(context)
    }

    @SuppressLint("DefaultLocale")
    private fun toonStatistiekWeerType(stats: ArrayList<StatistiekWeertype>) {

        val tvWeertype = findViewById<TextView>(R.id.tvWeertype)
        val tvGeenGegevens = findViewById<TextView>(R.id.tvGeenGegevens)

        if (mAllesJaarMaand === Helper.Periode.Alles) {
            tvWeertype.text = String.format("%s", getString(R.string.PerWeerType))
        }

        if (mAllesJaarMaand === Helper.Periode.Jaar) {
            tvWeertype.text = String.format("%s %s", getString(R.string.PerWeerType), String.format(getString(R.string.Jaartal), Integer.toString(mJaar)))
        }

        if (mAllesJaarMaand === Helper.Periode.Maand) {
            val dat = DateTime(mJaar, mMaand, 1, 0, 0)
            val mnd = dat.toString("MMM", Locale.getDefault()).replace(".", "")
            tvWeertype.text = String.format("%s %s", getString(R.string.PerWeerType), String.format(getString(R.string.JaartalEnMaand), mnd, Integer.toString(mJaar)))
        }

        val colors = ArrayList<Int>()
        val dataT = ArrayList<PieEntry>()
        val legendEntries = ArrayList<LegendEntry>()

        for (i in stats.indices) {
            val perc = stats[i].percentage
            val rPerc = Math.round(perc)
            dataT.add(PieEntry(perc, if (rPerc <= 2) "" else String.format("%d%%", rPerc)))
            val le = LegendEntry()
            val col = WeerHelper.weerTypeToWeerKleur(this, stats[i].weerType!!)
            colors.add(col)

            le.formColor = col
            le.label = stats[i].weerTypeOmschrijving
            le.form = Legend.LegendForm.SQUARE
            le.formSize = 10f
            legendEntries.add(le)
        }

        val chart = findViewById<PieChart>(R.id.pcPerWeertype)

        val desc = Description()
        desc.text = ""
        chart.description = desc

        chart.setTouchEnabled(false)
        chart.setNoDataText(getString(R.string.nodata))
        chart.legend.isEnabled = false
        chart.setEntryLabelColor(ContextCompat.getColor(this, R.color.colorPrimaryDark))
        chart.setDrawEntryLabels(true)

        val legend = chart.legend
        legend.setCustom(legendEntries)
        legend.isEnabled = true
        legend.xEntrySpace = 20f
        legend.textSize = 12f
        legend.isWordWrapEnabled = true

        val dsT = PieDataSet(dataT, "")
        dsT.colors = colors
        dsT.setDrawValues(false)

        val data = PieData(dsT)
        data.setValueTextSize(14f)
        data.setValueTextColor(ContextCompat.getColor(this, R.color.colorTekst))

        chart.data = data
        chart.animateXY(500, 500)
        if (stats.size == 0) {
            chart.visibility = View.GONE
            tvGeenGegevens.visibility = View.VISIBLE
        } else {
            tvGeenGegevens.visibility = View.GONE
            chart.visibility = View.VISIBLE
        }
        chart.invalidate()

    }

    private class AsyncGetStatistiekWeerTypeTask internal constructor(context: StatsWeerTypeActivity) : AsyncTask<Context, Void, ArrayList<StatistiekWeertype>>() {
        private val activityWeakReference: WeakReference<StatsWeerTypeActivity> = WeakReference(context)

        override fun doInBackground(vararg params: Context): ArrayList<StatistiekWeertype> {
            val context = params[0]
            val dh = DatabaseHelper.getInstance(context)
            return dh.getStatistiekWeerType(mAllesJaarMaand, mJaar, mMaand)
        }

        override fun onPostExecute(stats: ArrayList<StatistiekWeertype>) {
            val activity = activityWeakReference.get()
            activity?.toonStatistiekWeerType(stats)
        }
    }

    companion object {
        internal var mJaar = DateTime.now().year
        internal var mMaand = DateTime.now().monthOfYear
        internal var mAllesJaarMaand: Helper.Periode = Helper.Periode.Maand
    }
}